### DISCUSSION TITLE
ANSI escape codes in Julia output - again

### DISCUSSION THREAD

**Discussion by MHellmund** [quarto-dev/quarto-cli#9670] [Feature Requests]
### Description

Hi,
I wrote and use [a Quarto extension](https://github.com/MHellmund/julia-color)   (pandoc lua filter) handling ANSI colors (including RGB colors) and font attributes (bold, italic) for HTML, LaTeX and typst output. It needs a patched Quarto which just lets the ANSI codes  pass through to pandoc. My patch which adds a `preserve-ansi` option is [here](https://github.com/MHellmund/quarto-cli/commit/3fe8cab18a737c85c95ab4b2e9ae9fb17ac93f09). 

Take this as a cautious request. Would it be conceivable to integrate something like this option into Quarto? 

Many thanks for a great piece of software!

**Comment by mcanouil** [comment]
Did you try your extension using the new native Julia engine? https://quarto.org/docs/computations/julia.html#using-the-julia-engine

I am not sure Quarto does something to remove ANSI at least outside of Jupyter.

**Reply by mcanouil** [reply]
For reference:

- https://github.com/quarto-dev/quarto-cli/discussions/2630

**Reply by cscheid** [reply]
We're about to ship something similar with Typst (but not LaTeX), so hold tight!

**Reply by MHellmund** [reply]
@mcanouil : The julia engine uses the `jupyterToMarkdown` function of the jupyter engine where all the ansi code treatment/removal happens: https://github.com/quarto-dev/quarto-cli/blob/main/src/execute/julia.ts#L17 . So in this respect, it is the same. My quarto extension works, too, with the julia engine. 

@cscheid : That's fantastic. What's the plan? Invoking ansi_up and the new pandoc feature of translating css classes to typst? Does  it work for complex cases  like `class="ansi-green inverted" style="font-weight: bold; font-style: italic;"`?

Could you say something about when you will support typst output for quarto book projects?  Then I could put my patches to rest. Looking forward to it.

**Reply by gordonwoodhull** [reply]
Yes, we have just merged the part to translate CSS to Typst. Not all properties and elements are supported but it will grow and can be modified with Lua filters.

E.g. it has background and foreground color and `font-family` and `font-size` but not `font-weight` or `font-style`... yet. 

Those could be easily added in an additional Lua filter. See the existing filter here:
https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/filters/quarto-post/typst-css-to-props.lua

**Reply by gordonwoodhull** [reply]
My understanding is that we can almost support Typst books but there aren't a lot of Typst book templates available yet.

**Reply by MHellmund** [reply]
Thank you for all the good news!
